src/pybind/mgr/dashboard/controllers/osd.py

```diff
@@ -198,22 +205,25 @@ def _osd_flags():
     def list(self):
         return self._osd_flags()
 
-    def bulk_set(self, flags):
-        """
-        The `recovery_deletes`, `sortbitwise` and `pglog_hardlimit` flags cannot be unset.
-        `purged_snapshots` cannot even be set. It is therefore required to at
-        least include those four flags for a successful operation.
-        """
+    def bulk_delete(self, flags):
         assert isinstance(flags, list)
 
-        enabled_flags = set(self._osd_flags())
```

Comment:
By trying to be consistent, this introduced some possible issues. We should force the API consumer to state all possible flags and their status. Otherwise we might run into ACID issues.

Reply:
I don't understand how this can break ACID. Calling commands to Ceph can be atomic, yes, but not necessarily. So ACID is, kinda, thrown out of the window from the get go.
They are, however, idempotent. Even from a client's point of view, calling a command is not guaranteed to not having side-effects if it fails (think, for instance, a timeout on the client side). Side effects can still happen even if the command, from the client's point of view, failed. What we guarantee (to the best of our abilities) is that, should the *same* client run the command twice, without other clients issuing conflicting commands, the client will get the same result. E.g.,

> ceph osd pool set noin

and having it timing out because _reasons_ from the client's perspective, but having succeed on the cluster's side, will succeed as well when the client reruns the command.

I don't think we need, or should, require the client to call all possible flags. I think that will be hard to maintain as we add more flags, or as we change the way flags are consumed (or which flags are consumed). And, given we just want to modify a subset of the flags, sending all the flags regardless feels a bit... overengineered. Is it feasible from the cluster's point of view? Sure. Will those extra flags have unexpected results? Probably not. Unless they conflict with other concurrent operations, but I don't think that's something we should be concerned about, because if we're facing concurrent operations our state should be considered invalid from the start.

Reply:
You're right, this does not greatly modify the possible ACID issues (as is, the endpoint was retrieving the flag status in order not re-set all of them, and there a few possible race conditions there), I was rather thinking how the whole thing works from the end-user UI/UX.

As this currently is:
- The operator clicks on the "Cluster-wide OSD flags" button.
- They get a modal window with the current status of flags (which is not live updated). Let's assume `noin` is unticked.
- Another user either from CLI or dashboard sets `noin`.
- The first user sets `noout` and clicks submit (with `noin` unset).
- The resulting scenario is that `noin` is unset and `noout` is set.

I'm not sure if that's the best/desired scenario, just the one I tested. One workaround could be to make the OSD flags modal window refresh every x seconds.

Another approach could be expose/handle flags individually.